lib/DXIL/DxilUtil.cpp

```diff
@@ -608,64 +608,93 @@ bool IsResourceSingleComponent(Type *Ty) {
 }
 
 bool IsHLSLResourceType(llvm::Type *Ty) {
+  return std::get<0>(GetHLSLResourceType(Ty));
+}
+
+std::tuple<bool, hlsl::DXIL::ResourceClass, hlsl::DXIL::ResourceKind> GetHLSLResourceType(llvm::Type *Ty)
+{
+   using RetType = std::tuple<bool, hlsl::DXIL::ResourceClass, hlsl::DXIL::ResourceKind>;
+   RetType FalseRet(false, hlsl::DXIL::ResourceClass::Invalid, hlsl::DXIL::ResourceKind::Invalid);
+
   if (llvm::StructType *ST = dyn_cast<llvm::StructType>(Ty)) {
     if (!ST->hasName())
-      return false;
+      return FalseRet;
+
     StringRef name = ST->getName();
     ConsumePrefix(name, "class.");
     ConsumePrefix(name, "struct.");
 
     if (name == "SamplerState")
-      return true;
+      return RetType(true, hlsl::DXIL::ResourceClass::Sampler, hlsl::DXIL::ResourceKind::Sampler);
+
     if (name == "SamplerComparisonState")
-      return true;
+      return RetType(true, hlsl::DXIL::ResourceClass::Sampler, hlsl::DXIL::ResourceKind::SamplerComparison);
 
     if (name.startswith("AppendStructuredBuffer<"))
-      return true;
+      return RetType(true, hlsl::DXIL::ResourceClass::UAV, hlsl::DXIL::ResourceKind::StructuredBufferWithCounter);
+
     if (name.startswith("ConsumeStructuredBuffer<"))
-      return true;
+      return RetType(true, hlsl::DXIL::ResourceClass::UAV, hlsl::DXIL::ResourceKind::StructuredBufferWithCounter);
```

Comment:
SamplerComparison and StructuredBufferWithCounter were replaced with a property (SamplerCmpOrHasCounter) as part of SM6.6. 
https://github.com/microsoft/DirectXShaderCompiler/blob/cc43bc82b9a502b4a3828edc0bfea00ebc846a21/include/dxc/DXIL/DxilResourceProperties.h#L46